otherlibs/systhreads/st_stubs.c

```diff
@@ -376,13 +374,16 @@ static void caml_thread_initialize_domain()
   new_thread =
     (caml_thread_t) caml_stat_alloc_noexc(sizeof(struct caml_thread_struct));
 
+  if (new_thread == NULL)
+    caml_raise_out_of_memory();
```

Comment:
I don't think you can raise when `caml_thread_initialize_domain` is called from `caml_domain_start_hook`.
Or is there an argument where that is safe?

Reply:
If it was safe, we could just use `caml_stat_alloc`, which is equivalent to `caml_stat_alloc_noexc` + call to `caml_raise_out_of_memory` if NULL...

I'm not convinced it is safe to raise exceptions here.  This said, the OCaml 4 implementation of systhreads calls `caml_stat_alloc` (which can raise) in similar places in st_stubs.c, so it's fishy in OCaml 4 too.

Reply:
> quite dense

It is.

It is reasonable to introduce a “small allocation rule” that says that below a certain size, it is assumed that an allocation never fails, and allow the program to crash if it does. (Any allocation from OCaml small enough to go on the minor heap (<256 words) can crash the program anyway if it triggers a GC that in turn tries and fails to extend the major heap, and this is reportedly hard to fix.)

Having looked closely at the problem of asynchronous exceptions in the runtime, it is best if the OOM exception is reserved for large (implicit or explicit) allocations (or whose size depends on user input), and turned into a synchronous exception. (The runtime does not subscribe by this rule currently, but it also has a lot of theoretical resource leaks due to the OOM exceptions that are going to be very difficult to fix otherwise.)

(Users who want to limit their memory usage can do so differently, such as subscribing to an asynchronous event signalling memory pressure using cgroups, and converting this event into a proper asynchronous exception, which has the advantage to also work on overcommitting systems.)

`caml_fatal_error` is used for OOM in other places but since it now does `abort()` by default, which was reported not to be the best way to react to an allocation failure. But that's a separate problem.

So yes, it is reasonable to call `caml_fatal_error` here.

Reply:
I'm not happy with the caml_fatal_error solution, as I think we can report a proper `Out_of_memory` exception in this case.  I'll explain my idea next.

Reply:
Now that it is safe to raise from `caml_thread_initialize_domain`, you can shorten this to
```
  new_thread =
    (caml_thread_t) caml_stat_alloc(sizeof(struct caml_thread_struct));
```